### ISSUE TITLE
BUG?: precision issue using np.matmul on macOS 14 with Accelerate

### ISSUE THREAD

**Issue by jorisvandenbossche** [numpy/numpy#26791]
### Describe the issue:

I don't know if this is a bug (and if it would be, also probably in Accelerate and not NumPy) or rather a wrong expectation about floating point issues in matrix multiplication. 
But, for the downstream issue in shapely (https://github.com/shapely/shapely/issues/2066) this behaviour change _is_ a problem.

The issue is that we are transformating an array of coordinates using `np.matmul`, where the first and last coordinates are equal. After transforming them, the new first and last coordinates are still equal in general, except that this started to no longer be the case with NumPy 2.0 on MacOS 14 (so probably related to those wheels using Accelerate instead of OpenBLAS).

### Reproduce the code example:

```python
# using frombuffer to get the exact numbers at full precision
>>> coords = np.frombuffer(b'\xcb\x92\x87\x89\xfe\xde4@\x92\xce\xf0W\\L,\xc0\xb3\x03b\xffN!3\xc0\xe1y\x97b\x82\xb6+\xc0&ww\x99E\xe52\xc0\xfa\x8e\xbaY\xf7*0@\x04d\x85\xa0|\x1a5@\x14\x14h\xaap\xac/@\xcb\x92\x87\x89\xfe\xde4@\x92\xce\xf0W\\L,\xc0').reshape(5, 2)
>>> coords
array([[ 20.87107143, -14.14914203],
       [-19.13011166, -13.85646351],
       [-18.89559325,  16.16783677],
       [ 21.10346416,  15.83679707],
       [ 20.87107143, -14.14914203]])

>>> A = np.frombuffer(b'7\xb3\xf1\xdd\xc7\xff\xef\xbf\xeb\xe4b\x9b\xf5\xf7}?\xeb\xe4b\x9b\xf5\xf7}\xbf7\xb3\xf1\xdd\xc7\xff\xef\xbf').reshape(2, 2)
>>> A
array([[-0.99997323,  0.00731655],
       [-0.00731655, -0.99997323]])

>>> result = np.matmul(A, coords.T).T
>>> result
array([[-20.9740357 ,  13.99605905],
       [ 19.0282181 ,  13.99605905],
       [ 19.01338028, -16.02915345],
       [-20.98702857, -15.99077774],
       [-20.9740357 ,  13.99605905]])

# first and last row have *exactly* equal values in the input
>>> coords[0, :] == coords[-1, :]
array([ True,  True])

# for me (on linux, with OpenBLAS) this is still the case in the output
>>> result[0, :] == result[-1, :]
array([ True,  True])
```

### Python and NumPy Versions:

2.0.0
3.12.2 | packaged by conda-forge | (main, Feb 16 2024, 20:50:58) [GCC 12.3.0]

**Comment by astrofrog** [comment]
To confirm, on MacOS 14 with the Numpy 2.0 wheel from PyPI:

```python
>>> print(coords[0, :] == coords[-1, :])
[ True  True]
>>> print(result[0, :] == result[-1, :])
[ True False]
```

**Comment by andyfaff** [comment]
```python
>>> np.testing.assert_equal(result[0, :], result[-1, :])
<snip>
AssertionError: 
Arrays are not equal

Mismatched elements: 1 [/](http://localhost:8888/) 2 (50%)
Max absolute difference among violations: 1.77635684e-15
Max relative difference among violations: 1.26918358e-16
 ACTUAL: array([-20.974036,  13.996059])
 DESIRED: array([-20.974036,  13.996059])

>>> print(np.show_runtime())
[{'numpy_version': '2.0.0',
  'python': '3.12.3 | packaged by conda-forge | (main, Apr 15 2024, 18:35:20) '
            '[Clang 16.0.6 ]',
  'uname': uname_result(system='Darwin', node='someipaddress, release='23.5.0', version='Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000', machine='arm64')},
 {'simd_extensions': {'baseline': ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD'],
                      'found': ['ASIMDHP'],
                      'not_found': ['ASIMDFHM']}}]
None
```

**Comment by rgommers** [comment]
> Should we generally not rely on the `np.matmul` returning identical values for identical values in the input array?

I'd be inclined to say that relying on exact equality for numerical functions is technically always a bug, unless it's explicitly documented/guaranteed in the function that you can do so. 

I don't know why the results are off by 1 ULP in this case, but I'm not particularly surprised by it.

**Comment by jorisvandenbossche** [comment]
> I'd be inclined to say that relying on exact equality for numerical functions is technically always a bug, unless it's explicitly documented/guaranteed in the function that you can do so.

And is that the case for plain addition/multiplication? 
On the shapely side, before using `np.matmul` we were calculating the same with a written out formula with just some additions and multiplications (something like `x = coords[:, 0]; y = coords[:, 1]; xp = a * x + b * y + xoff; yp = d * x + e * y + yoff` and then stacking `xp` and `yp` to get the resulting array)

**Comment by rgommers** [comment]
> And is that the case for plain addition/multiplication?

Yes, still the case then. Like for `matmul` I'd say it's highly likely that you'll get identical results, but not 100% guaranteed. Depending on array sizes, compilers, SIMD code under the hood etc., results for some values could be off by 1 ULP.

**Comment by ngoldbaum** [comment]
Floating point addition and multiplication aren't associative, so if for some reason the compiler re-orders the operations you may get a different result out of the computation.

**Comment by seberg** [comment]
I guess the question is why it should change in result and that would be strides or alignment and thus either use of different registers or different iteration order alltogether?
Could be a fused multiply-add, but not for the peeled end or so?

I think that the NumPy version should be reproducible.  It only uses normal addition and multiplication, which should be exact always.  There cannot be any reordering (which could matter in higher dimensions) since you hardcode the order, and it cannot contract to fused multiply-add based on something random alignment or so, since NumPy doesn't do magic.
Or am I missing a way that this might go wrong for simple addition/multiplication intrinsics!?

Now if it was implemented using a `.sum()` then memory order could make a difference (but for NumPy not if it is in the same array, I think).

**Comment by rgommers** [comment]
> since NumPy doesn't do magic.

Temporary elision is slightly magical:) It should happen for both statements since arrays are of the same size, but it's already not a simple addition anymore, it switches from out-of-place to in-place. I think you're right though that it'll still yield identical numbers - but it's not something numpy guarantees. This is a fun academic debate, but it's best to just change the downstream code.

**Comment by jorisvandenbossche** [comment]
Thanks for all the feedback! 
We should consider then how to handle this on the shapely side (ideally we make it robust for such small differences of 1 ULP when recreating the Polygon from the transformed coordinates, but that's something for which we would need to add custom code as the underlying GEOS library does not do that). Although short term, my feeling is that going back to the plain numpy multiplication/addition is probably fine for now (that has worked for many years without problem, although I understand numpy gets smarter and with more recent versions this might also not stay true)

**Comment by rkern** [comment]
I find that with computational geometry code, it usually pays to be explicit and granular and avoid higher-level abstractions (from third parties, at least). Computational geometry is rife with invariants that need to be preserved that our typical abstractions for floating point computations rarely preserve.

In this case, you might consider using `matmul()` and fixing up the last point by copying in the result for the first point. You don't need to rely on `matmul()` preserving the invariant since this invariant is luckily relatively trivial to simply enforce at any time.
